Resolve deprecation warnings for v0.7

halleysfifthinc opened the pull request:
According to the testsuite, this fixes all the deprecations.

The only things I'm unsure of is the change in v0.7 for reinterpret/reshape to return `reinterpret`/`reshape` types. Since the tests check type equality (and error if mismatched type) before checking for equality in the contents, this required me to add a `collect` (for most of the cases) around the reinterpret or reshape. This will now allocate when it did not before, and may affect the read/write performance, however I don't see a way around this (other than changing the types of the test data to make the tests pass with reinterpret/reshape types).

halleysfifthinc commented:
Tests fail due to HDF5 failing to install/build.

musm commented on src/MAT_HDF5.jl:
```diff
@@ -28,15 +28,15 @@
 const HDF5BitsOrBool = Union{HDF5BitsKind,Bool}
 
-type MatlabHDF5File <: HDF5.DataFile
+mutable struct MatlabHDF5File <: HDF5.DataFile
```
I think this can simple be a struct

halleysfifthinc replied:
Ok. Do you know if the other one or two occurrences of mutable structs can be constant as well?

musm replied:
I believe they should all be struct's

halleysfifthinc replied:
Nope. Turns out both the `MatlabHDF5File` and `Matlabv5File` need to be mutable.

Evizero commented:
What is the state on this?

halleysfifthinc commented:
My understanding is that the PR is ready to be merged, depending on how important passing CI is for whoever has merge permissions. Tests pass locally on my machine (Linux x64), but the last run of Travis failed on 1.0 (bad travis config for 1.0, no `using Pkg`), tests on 0.7 failed because CMakeWrapper failed to build, Blosc failed because of CMakeWrapper. HDF5 failed to install because the apt sources returned 404's; my guess is that is a Travis issue, and I have no clue if there is anything we can do to solve that on our end.

musm commented on REQUIRE:
```diff
@@ -1,4 +1,4 @@
-julia 0.5
+julia 0.7-beta2
```
just `julia 0.7`

musm commented on src/MAT_HDF5.jl:
```diff
@@ -290,12 +290,12 @@ elseif length(s) > 63
 
 toarray(x::Array) = x
-toarray(x::Array{Bool}) = reinterpret(UInt8, x)
+toarray(x::Array{Bool}) = reinterpret.(UInt8, x)
```
I think you can delete this
`toarray.(x)` should work if  `toarray(a)` works, where `a` is a Bool, be the definition below.

halleysfifthinc replied:
It's necessary for correct matching during multiple dispatch.

halleysfifthinc commented:
All green! @simonster or @yuyichao could we get this merged now?

yuyichao commented:
> Since the tests check type equality (and error if mismatched type) before checking for equality in the contents, this required me to add a collect (for most of the cases) around the reinterpret or reshape. This will now allocate when it did not before, and may affect the read/write performance, however I don't see a way around this (other than changing the types of the test data to make the tests pass with reinterpret/reshape types).

Most of those looks wrong. You should now allocate the array you want to return and fill the reinterpreted/reshaped version.

halleysfifthinc commented:
Ok. I will see about changing that.

yuyichao commented on test/write.jl:
```diff
@@ -33,10 +33,10 @@ test_write(Dict(
 ))
 
 test_write(Dict(
-	"Complex128" => [1.0 -1.0 1.0+1.0im 1.0-1.0im -1.0+1.0im -1.0-1.0im 1.0im],
```
https://github.com/JuliaIO/MAT.jl/pull/86

yuyichao commented on test/write.jl:
```diff
@@ -33,10 +33,10 @@ test_write(Dict(
 ))
 
 test_write(Dict(
-	"Complex128" => [1.0 -1.0 1.0+1.0im 1.0-1.0im -1.0+1.0im -1.0-1.0im 1.0im],
```
Changing this seems OK but if the test fails without changing this then something is wrong.

halleysfifthinc replied:
No that was just a name change to stay current with names and such. Tests pass with or without.

bjarthur commented:
looking forward to having this!

i think it would be okay for MAT.jl to return `ReinterpretArray`'s.  so no need for a `collect` there.  but `h5d_write` probably cannot deal with those, so a `collect` might be needed when writing.  i presume though that the previous behavior of `reinterpret` did a collect internally, so the performance might be similar for writes and better for reads.  thoughts @yuyichao ?

halleysfifthinc commented:
While I don't think that returning `ReinterpretArray`s would be ok, I do agree with you regarding the performance of `reinterpret`/`collect` vs. Julia v0.6. My intuition/understanding is that is how reinterpret previously worked, so putting a `collect` on the reinterpreted arrays should give at least equivalent performance.

At the same time, I share @yuyichao sense of wrongness about collecting reinterpreted arrays. However, in looking at the first couple of occurrences of `reinterpret`, I'm not sure how to cleanly implement his suggestion without dramatically restructuring the code, and I'm not comfortable enough with this code to do that.

Perhaps we could merge and tag this now, and then the `collect`s and `reinterpret`s could be dealt with later? There seem to be a decent number of people waiting for MAT to be updated.

bjarthur commented:
i get this warning when testing this PR on julia 0.7 on linux:

```
┌ Warning: Vector{UInt8}(s::String) will copy data in the future. To avoid copying, use `unsafe_wrap` or `codeunits` instead.
│   caller = close(::MAT.MAT_HDF5.MatlabHDF5File) at MAT_HDF5.jl:65
└ @ MAT.MAT_HDF5 ~/.julia/dev/MAT/src/MAT_HDF5.jl:65
```

halleysfifthinc commented:
I just fixed that warning and removed nearly all of the reinterprets. I'm not sure why v1 is failing on OSX (but passing on v0.7 and nightly), also, the logs on the passing builds were incomplete and still showed as passing. I'll see if rerunning Travis changes anything.

halleysfifthinc commented:
Anyone have a mac and could investigate this? I'm on Ubuntu 18.04 x86_64 with hdf5-tools at version 1.10.0.

Travis seems a bit wonky today. The logs are delayed and/or truncated, so I can't look to see if there is any difference between the passing and not passing OSX runs.

bjarthur commented:
tests pass locally for me on OSX for julia 1.0.

@yuyichao any further comments?

halleysfifthinc commented on src/MAT_HDF5.jl:
```diff
@@ -336,15 +330,14 @@ function m_writearray{T<:HDF5BitsOrBool}(parent::HDF5Parent, name::String, adata
         try
-            arr = reinterpret(T, adata, tuple(2, size(adata)...))
-            HDF5.writearray(dset, dtype.id, arr)
+            HDF5.writearray(dset, dtype.id, vec(adata))
```
It might (I don't recall) but the more relevant point is that it shouldn't be necessary. Correct me if I'm wrong, but I don't think it matters to HDF5 what the originating type is, only what `dtype.id` says it is?

yuyichao commented on src/MAT_HDF5.jl:
```diff
@@ -290,12 +284,12 @@ elseif length(s) > 63
 
 toarray(x::Array) = x
-toarray(x::Array{Bool}) = reinterpret(UInt8, x)
+toarray(x::Array{Bool}) = reinterpret.(UInt8, x)
```
I mean this is a copy that didn't exist before.

halleysfifthinc replied:
Just to be clear, you are referring to the broadcasted reinterpret?

nicoleepp replied:
Can this be solved by simply doing `toarray(x::Array{Bool}) = Array{UInt8}(x)`?

yuyichao replied:
No that s copying

halleysfifthinc commented:
Regarding the last two places with the extra copies, I'm not sure how avoid them without returning a `reinterpret` type array. I recognize that @yuyichao doesn't have the time right to look into them more, perhaps @bjarthur has some suggestions? Otherwise this can wait until someone with more experience with Julia can take a look.

yuyichao commented:
And the reason I believe they are all fixable without any extra copy is that I believe all what the C library (or HDF5 format/package) should care about is a memory buffer to accept data/fill data in. The type tag on it should not matter so allocating the correct final type and let the library fill the data in should always work.

This might not be compatible with the existing internal or HDF5 API and so fixing this properly might require a non-local change involving internal or external (API) change/addition, which is what I don't really have time to investigate now...

halleysfifthinc commented:
That is my sense of things as well. Given that Travis passes now, could this be merged and the extra copies be opened as an issue?

yuyichao commented:
I strongly perfer to NOT do that. There should be no rush to get everything working on 1.0 since

1. 0.6 still works.
2. I don't believe anyone will actually fix the extra copy if it was not fixed now. It'll be very hard to just figure out which copy is "extra" after this is merged.

musm commented on src/MAT_HDF5.jl:
```diff
@@ -598,7 +598,9 @@ function read(obj::HDF5Object, ::Type{Bool})
-    Array{Bool}(tf .> 0)
+    r = Array{Bool}(undef, size(tf))
+    r .= tf .> 0
+    return r
```
why this change? AFAIK `reinterpret(Bool,...)` works ?

c-p-murphy commented:
I took a quick look at this, but I'm in a little over my head. Hopefully this will help out though. 

1) It looks like a reinterpreted array isn't a subtype [here](https://github.com/JuliaIO/MAT.jl/blob/a1e7ae622f5d22f71c7253a87c3cb76b96cc0345/src/MAT_HDF5.jl#L323I).
2) When that is resolved, the next issue is [here](https://github.com/c-p-murphy/HDF5.jl/blob/master/src/HDF5.jl#L1174).
3) And I think part of 2. is that there needs to be something added around [here](https://github.com/c-p-murphy/HDF5.jl/blob/master/src/HDF5.jl#L268-L282).

nicoleepp commented:
https://github.com/JuliaIO/HDF5.jl/pull/520 should make `reinterpret` work where there currently are extra copies.

yuyichao commented on src/MAT_HDF5.jl:
```diff
@@ -62,7 +62,11 @@ function close(f::MatlabHDF5File)
             magic = zeros(UInt8, 512)
             identifier = "MATLAB 7.3 MAT-file" # minimal but sufficient
-            magic[1:length(identifier)] = Vector{UInt8}(identifier)
+            GC.@preserve magic identifier begin
```
This has nothing to do with this PR so you don't have to change anything but I feel like `magic` should just be a global constant array here...

bjarthur commented:
given https://github.com/JuliaIO/HDF5.jl/pull/520, does anything need to change in this PR still?  or are we go for merge?

yuyichao commented:
Yes, a few copy still needs to be removed here.

zjroth commented:
Would changing
```julia
toarray(x::Array{Bool}) = reinterpret.(UInt8, x)
```
to
```julia
function toarray(x::Array{Bool})
    x_uint8 = reinterpret(UInt8, x)
    unsafe_wrap(Array{UInt8}, pointer(x_uint8), size(x_uint8))
end
```
get rid of the copy in question in an acceptable way?  It seems to work on my system.

yuyichao commented:
No that's illegal

zjroth commented:
> No that's illegal

Okay.  Sorry for the noise.

th85ac commented:
Not running for me under Julia 1.0.1 (mac os x)... all other packages like HDF5 working

oxinabox commented:
> ... Given that Travis passes now, could this be merged and the extra copies be opened as an issue?

re:

> I strongly perfer to NOT do that. There should be no rush to get everything working on 1.0 since
>
> 1.    0.6 still works.
> 2.    I don't believe anyone will actually fix the extra copy if it was not fixed now. It'll be very hard to just figure out which copy is "extra" after this is merged.



@yuyichao   is this still your strong perference?

Re point 1.
Given now a lot of us have migrated fully off 0.6 (or potentially are being blocked from migration by this)

Obviously this will not effect 0.6 users since this is tagged 0.7+


Re point 2.
Has the extra copies been fixed now in the last set of changes?
I see some are tagged as fixing extra copies.

timholy commented:
@yuyichao, I greatly appreciate the time you're taking to review the various PRs here, but I'd like to put in a plug for getting this resolved in a way that doesn't turn the perfect into the enemy of the good.  We know of many imperfections throughout Julia itself and in many other packages, yet often we soldier on, doing the best we can at the time. There are times where an imperfect solution would actively undermine progress, but if the issue here is simply making some extra copies then this doesn't seem to be in that category. (It doesn't have exported API implications, for example, right?)

Unlike you, I haven't taken the time to review the various PRs, so I haven't followed the underlying issues. Would it be reasonable to merge something that at least works and then file an issue detailing the deficiencies that need fixing?

yuyichao commented:
I believe the copying issue is only needed to maintain the output type which shouldn't break most of the use so merging PRs that doesn't include that part is certainly fine and I've done that already.

oxinabox commented:
Just to be clear *master* is still broken in 1.0
because a ton of the basic deprecatation removing stuff has not been done (because it is done in this PR).

And *master* is also broken on  0.7 is also broken because of the reinterpretted output types thing.

halleysfifthinc commented:
I noticed the same thing about the updates to master; it was unusable for
me (as of a week ago).

I haven't had the time recently to take a look at this again, but the brief
look I took at the PR to HDF5 linked earlier in this thread looked to be on
the right track. I may have some time in the next week or two to take
another stab at this.

On Fri, Nov 16, 2018, 10:30 AM Lyndon White <notifications@github.com>
wrote:

> Just to be clear *master* is still broken in 1.0
> because a ton of the basic deprecatation removing stuff has not been done
> (because it is done in this PR).
>
> And *master* is also broken on 0.7 is also broken because of the
> reinterpretted output types thing.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaIO/MAT.jl/pull/86#issuecomment-439429605>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHA_LfV4pYyWJaf6515Cq70slOMckWqVks5uvtoygaJpZM4VnKUN>
> .
>
-- 
-AH

nicoleepp commented:
The HDF5 changes needed to fix the extra copies here haven't been tagged yet:
https://github.com/JuliaIO/HDF5.jl/issues/535

oxinabox commented:
Here is the METADATA PR.
It is currently kinda stalled
https://github.com/JuliaLang/METADATA.jl/pull/19579

oxinabox commented:
The HDF5 changes are merged and tagged. Thanks @musm

halleysfifthinc commented:
I think that should fix all the extra copies. If somebody (@yuyichao perhaps) can review to double-check, then this will be good to go once JuliaIO/HDF5.jl#538 is merged and tagged.

yuyichao commented:
See https://github.com/JuliaIO/HDF5.jl/pull/538#issuecomment-445867472

davidssmith commented:
Can we get this merged? 

Do you know what is worse than a performance penalty? A broken master. That's infinitely bad performance.

yuyichao commented:
Not yet (see comment) and no. Old julia version aren't gone yet.

Also note that master isn't completely broken (some cases works) and I've said many times that whoever want to split out part of this to another PR are welcome to do so. I've merged such PR and I've rebased this on that.

Also note that this version is also "broken" awaiting the HDF5 PR linked above.

davidssmith commented:
Well it is broken for me, and I created a PR a long time ago that made it work for me, but it was rejected for the reason that it was incomplete, even though it fixed some things.

Don't let perfect be the enemy of the good.

yuyichao commented:
I am very surprised. If you are talking about https://github.com/JuliaIO/MAT.jl/pull/92. I believe it's included in https://github.com/JuliaIO/MAT.jl/pull/100, if not, feel free to include the missing pieces.
Not being complete is not an issue. https://github.com/JuliaIO/MAT.jl/pull/100 is just like that. However, if it does not completely fix all issues on 1.0, it should not break 0.6.

davidssmith commented:
> There should be no rush to get everything working on 1.0 since

Also I want to directly disagree with this. Many packages depend on this as part of their backend. The performance of reading and writing MAT files is not as important as allowing all of those packages to upgrade to 1.0+, where they may enjoy speed improvements in execution and better likelihood of adoption. As it is now, anyone who hears about Julia, goes to download 1.0, and tries to install MAT or some package depending on MAT, will get errors. That's not the way to win hearts and minds.

yuyichao commented:
You are just giving very good reason for why the release 1.0 shouldn't have been made so quickly. 0.6 is still a perfectly good release to use and start with. I know this is directly conflicting with the super loose release policy and I don't want to argue with either.

Also, I'm not looking for perfect either. It's just about not introducing regression. And if you want to know what is the right way to do this on a very much related change? https://github.com/JuliaLang/julia/pull/21831#issuecomment-302180760 Noticed that the PRs are submitted to the packages to make sure there's no regression **before** the base PR is merged.

oxinabox commented:
> However, if it does not completely fix all issues on 1.0, it should not break 0.6.

I must be missing something.
This PR will change the minimum julia version to be 0.7
It thus **cannot** break julia 0.6.


If you are saying that you don't want to lose the ability to make improvements to 0.6,
then have a backports branch.
I am maintaining 2 packages that have backports branchs and it isn't that hard.

babaq commented:
is this branch dead or not, I think it's better get merged and make it work first, and then make another PR to fix performance regression. v1.0 is relatively big change to v0.6 that's why it's better to move to v1.0 early than doing more work in v0.6. please make this work early, cause a lot of data is in MAT format.

chriselrod commented:
FWIW, this branch works for me, but the merged one errors.

timholy commented:
Superseded by #110 (merged in #113)
